### ISSUE TITLE
Refactoring idea - conf vs state

### ISSUE THREAD

**Issue by kum-deepak** [dc-js/dc.js#1707]
The dc classes currently have quite a lot of private fields. I can see a particular type of field - these are assigned by the calling code, typically has a default, however, never updated by dc itself. I was thinking of moving these into an object `conf` (key/value pairs). So effectively `this._transitionDuration` will become `this.conf.transitionDuration`. Some additional notes:

- This will help in separating out chart state from the configuration. 
- Some classes (like `ColorMixin`) may end up having no state - this will help in further refactoring that class.
- We will define a set of interfaces (not every chart will need the same) for the conf classes. This will allow type checking for the library and the typescript users.
- An interesting benefit will also come. Typically there are several charts on a single page. A common conf object can be passed by the user to several charts.

This whole idea came while trying to figure out which fields represent the chart state.

**Comment by gordonwoodhull** [comment]
I agree - it is frustrating to set the same parameters on everything. Particularly for things that must be the same everywhere like `commitHandler`

Although, this won’t help if two charts need *mostly* the same parameters, except for one thing...

**Comment by kum-deepak** [comment]
Actually with ES6 that is not that difficult, bit clumsy maybe:

```typescript
conf2 = Object.assign({}, conf1, {transitionDuration: 300});
```

It has been used in https://github.com/dc-js/dc.js/blob/develop/rollup.config.js#L45 as well.

**Comment by gordonwoodhull** [comment]
Oh you mean a plain object, not a class. Interesting - makes the getters/setters kind of superfluous but that should be fine.

That’s the ES5 way. The spread operator in ES6 is nicer and had the same effect:

```js
conf2 = {...conf1, transitionDuration: 300}
```

**Comment by kum-deepak** [comment]
Wow! I did not realize this syntax. :smile:

**Comment by kum-deepak** [comment]
Closed by #1736
